packages/@react-aria/color/src/useColorField.ts

```diff
@@ -71,24 +77,147 @@ export function useColorField(
+      // the undo/redo stack.
+      if (nextValue == null || !state.validate(nextValue)) {
+        e.preventDefault();
+      }
+    };
+
+    input.addEventListener('beforeinput', onBeforeInput, false);
+    return () => {
+      input.removeEventListener('beforeinput', onBeforeInput, false);
+    };
+  }, [ref, stateRef]);
+
+  let onBeforeInput = !supportsNativeBeforeInputEvent()
+    ? e => {
+      let nextValue =
+        e.target.value.slice(0, e.target.selectionStart) +
+        e.data +
+        e.target.value.slice(e.target.selectionEnd);
+
+      if (!state.validate(nextValue)) {
+        e.preventDefault();
+      }
+    }
+    : null;
+
+  let onChange = value => {
+    state.setInputValue(value);
   };
 
+  let compositionStartState = useRef(null);
+
   let {labelProps, inputProps} = useTextField(
     mergeProps(props, {
       id: inputId,
       value: inputValue,
       type: 'text',
       autoComplete: 'off',
-      onChange: setInputValue
+      onChange,
+      onBeforeInput,
+      onCompositionStart() {
+        // Chrome does not implement Input Events Level 2, which specifies the insertFromComposition
+        // and deleteByComposition inputType values for the beforeinput event. These are meant to occur
+        // at the end of a composition (e.g. Pinyin IME, Android auto correct, etc.), and crucially, are
+        // cancelable. The insertCompositionText and deleteCompositionText input types are not cancelable,
+        // nor would we want to cancel them because the input from the user is incomplete at that point.
+        // In Safari, insertFromComposition/deleteFromComposition will fire, however, allowing us to cancel
+        // the final composition result if it is invalid. As a fallback for Chrome and Firefox, which either
+        // don't support Input Events Level 2, or beforeinput at all, we store the state of the input when
+        // the compositionstart event fires, and undo the changes in compositionend (below) if it is invalid.
+        // Unfortunately, this messes up the undo/redo stack, but until insertFromComposition/deleteByComposition
+        // are implemented, there is no other way to prevent composed input.
+        // See https://bugs.chromium.org/p/chromium/issues/detail?id=1022204
+        let {value, selectionStart, selectionEnd} = ref.current;
+        compositionStartState.current = {value, selectionStart, selectionEnd};
+      },
+      onCompositionEnd() {
+        if (!state.validate(ref.current.value)) {
+          // Restore the input value in the DOM immediately so we can synchronously update the selection position.
+          // But also update the value in React state as well so it is correct for future updates.
+          let {value, selectionStart, selectionEnd} = compositionStartState.current;
+          ref.current.value = value;
+          ref.current.setSelectionRange(selectionStart, selectionEnd);
+          state.setInputValue(value);
+        }
+      }
```

Comment:
A lot of the changes in this file look like they are pretty much identical to NumberField. Should we have a shared hook for this? Like `useFormattedTextField` or something? Looks like basically all it needs is a `validate` method?